server/consumer.go

```diff
@@ -3650,7 +3650,8 @@ func (o *consumer) nextWaiting(sz int) *waitingRequest {
 					// If we have a match, we do nothing here and will deliver the message later down the code path.
 				} else if wr.priorityGroup.Id == _EMPTY_ {
 					o.waiting.cycle()
-					if wr == lastRequest {
+					numCycled++
+					if numCycled >= o.waiting.len() {
```

Comment:
@neilalexander I was wondering if it's possible for `o.waiting` to be modified with new requests during the execution of `nextWaiting`? I don't think so, as I see that everything is done under lock in `loopAndGatherMsgs` but I'd like to make sure comparing to `o.waiting.len()` in here is a valid approach.

Reply:
I think `o.waiting.len()` should be safe, AFAIK the only places where new requests are added are under the consumer lock.